src/IO/BufferWithOwnMemory.h

```diff
@@ -108,24 +112,47 @@ struct Memory : boost::noncopyable, Allocator
         if (!(value % alignment))
             return value;
 
-        return (value + alignment - 1) / alignment * alignment;
+        // original expression is (value + alignment - 1) / alignment * alignment;
+
+        size_t res = 0;
+
+        if (common::addOverflow<size_t>(value, alignment - 1, res))
+            throw Exception("value is too big to apply alignment", ErrorCodes::ARGUMENT_OUT_OF_BOUND);
```

Comment:
I've made this calculation save and covered it with tests.
But.
I think that code is redundant. We don't have to round size manually. All the work is done by Allocator. The allocator is responsible for ensuring that the result address is a multiple of the alignment. It is done with posix_memalign.
I want to try make implementation without function align at all. Left only padding.

Reply:
I guess the point was that you may allocate 1 byte with alignment to 16 bytes and have capacity 16. if you later call resize for 2, 3, 4, ... , 16 bytes it won't lead to reallocation

Reply:
I also think that the purpose was exactly that. It should be named round instead alignment.
I greped the code. Class Memory usually is using with default alignment=0. Except one (? may be several) case with alignment=pageSize. And this is special case when pages are pinned. Of course, I could be wrong, but at least I've really tried to check this and this is the result of it.